### PR TITLE
[AzureMonitorExporter] implementing Truncation rules for ExceptionDetails and StackFrame

### PR DESCRIPTION
**Pull request by TimothyMothra**
Continuation of #29094

## Changes
- new class `StringExtensions` to hold `Truncation()` extension method.
- new `partial class StackFrame`
  - custom ctor with truncation rules.
- `partial class TelemetryExceptionDetails`
  - remove use of base ctor. Need to apply truncation rules to props in the base ctor, have to implement this in custom ctor.
  - remove `public const int MaxParsedStackLength`, use `SchemaConstants` instead.
  - moved helper methods to `StackFrame` class.
  - refactor method `SanitizeStackFrame`.
    - old: `Tuple<List<TOutput>, bool> SanitizeStackFrame<TInput, TOutput>(IList<TInput> inputList, Func<TInput, int, TOutput> converter, Func<TOutput, int> lengthGetter)` 
    - new: `List<StackFrame> GetSanitizedStackFrame(Exception exception, out bool hasFullStack)`

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/StringExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        /// This method will truncate a string if that string exceeds a specified max length.
+        /// </summary>
+        /// <remarks>
+        /// This method is a wrapper around <see cref="string.Substring(int, int)"/>.
+        /// </remarks>
+        /// <param name="input">A string to be evaluated.</param>
+        /// <param name="maxLength">A specified length which is used to evaluate the input string.</param>
+        /// <returns>The input string if less than max length, or a substring that begins at 0.</returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static string Truncate(this string input, int maxLength)
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/StringExtensions.cs:20]: https://github.com/open-telemetry/opentelemetry-dotnet/issues/2919 Check this, and consider following that approach.

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/StringExtensions.cs:20]: I'm not sure what alternatives would be appropriate here. Do you have any ideas?

> As rule of thumb, I avoid adding extension methods for BCL classes unless these are very generic.
> https://github.com/open-telemetry/opentelemetry-dotnet/pull/2914#discussion_r809758538 

I think this `Truncate` method meets the bar of being very generic.
I've got almost 100 constants defined in [SchemaConstants](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/SchemaConstants.cs) and my intent is to use this extension method for all of these.

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/StringExtensions.cs:20]: Okay if it is generic enough, ( I agree), this is fine.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryExceptionDetails.cs [hunk 2]
```diff
@@ -4,124 +4,83 @@
...
+            if (frames != null && frames.Any())
             {
-                int currentParsedStackLength = 0;
-                for (int level = 0; level < inputList.Count; level++)
+                int totalLength = 0;
+                for (int counter = 0; counter < frames.Length; counter++)
                 {
-                    // Skip middle part of the stack
-                    int current = (level % 2 == 0) ? (inputList.Count - 1 - (level / 2)) : (level / 2);
+                    // 1) Check that we don't exceeded MaxFrames.
```
Comment (vishweshbankwar) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryExceptionDetails.cs:52]:
```suggestion
                    // Check if we have exceeded MaxFrames.
```

**Review by vishweshbankwar — APPROVED** [review]

**Review by rajkumar-rangaraj — APPROVED** [review]

**Pull request merged**